### PR TITLE
[5.0.x] SqlServer: Add Cast to (n)varchar(max) when injecting Concat for multi-line seed data

### LINKED ISSUE #24112: Using Concat for model/seed data in migrations may truncate long strings
Originally reported as https://github.com/dotnet/efcore/pull/23634#issuecomment-776769039 by @saliksaly.

Using CONCAT for strings is not so good due to string truncation to 4000 chars in SQL server :-(
Now the migrations run without error (Microsoft.Data.SqlClient.SqlException (0x80131904): The concat function requires 2 to 254 arguments.) but my seeded strings are now truncated in the database. Silently!

Here are some resources for this SQL server behavior:
https://stackoverflow.com/questions/4547821/using-varcharmax-with-string-concatenation-in-sql-server-2005
https://stackoverflow.com/questions/4833549/nvarcharmax-still-being-truncated 
https://sqlundercover.com/2017/09/27/concatenation-truncation-are-your-strings-being-truncated/

Is there any workaround for this? We have had CSHTML templates in seed data and now it would not be posible in new version of .net?

### PR DESCRIPTION
**Pull request by smitpatel**
Resolves #24112

**Description**

Silent truncation happens when using seed data containing line breaks and longer than 4000 bytes length.

**Customer Impact**

This causes corruption in user data due to truncation.

**How found**

Reported by a customer.

**Test coverage**

Test coverage for this case has been added in this PR.

**Regression?**

Yes, from EF Core 3.1.

**Risk**

Low. The change happens in SQL literal generation which is used extensively in various parts to generate SQL and well covered by tests. The fix is also quirked.

### CODE REVIEW

File: src/EFCore.SqlServer/Storage/Internal/SqlServerStringTypeMapping.cs [hunk 1]
```diff
@@ -298,6 +304,18 @@ void AddConcatOperatorIfNeeded()
...
                 {
+                    if (!useOldBehavior2
+                        && !castApplied)
+                    {
+                        builder.Append(" AS ");
+                        if (IsUnicode)
+                        {
+                            builder.Append("N");
+                        }
+                        builder.Append("VARCHAR(MAX))");
```
Comment (bricelam) [inline src/EFCore.SqlServer/Storage/Internal/SqlServerStringTypeMapping.cs:315]: Please don't shout such obscenities at our users... 😉
```suggestion
                            builder.Append("n");
                        }
                        builder.Append("varchar(max))");
```

**Review by bricelam — APPROVED** [review]

File: src/EFCore.SqlServer/Storage/Internal/SqlServerStringTypeMapping.cs (hunk 1 above)
Comment (smitpatel) [reply src/EFCore.SqlServer/Storage/Internal/SqlServerStringTypeMapping.cs:315]: So Query pipeline uses small case when writing store type name but migrations pipeline uses upper case. So where are we heading to?

Comment (bricelam) [reply src/EFCore.SqlServer/Storage/Internal/SqlServerStringTypeMapping.cs:315]: All SQL Server types should be lower-case. I tried to fix everything in 6.0…

Comment (smitpatel) [reply src/EFCore.SqlServer/Storage/Internal/SqlServerStringTypeMapping.cs:315]: So are we updating this in 5.0.x? I am fine either way.

Comment (bricelam) [reply src/EFCore.SqlServer/Storage/Internal/SqlServerStringTypeMapping.cs:315]: I just wanted any new 5.0 code to use lower-case so it’s eventually consistent when we merge it into 6.0

**Pull request merged**